core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthServiceBuilder.java

```diff
@@ -0,0 +1,135 @@
+        }
+
+        this.delegate = requireNonNull(delegate, "delegate");
+        return this;
+    }
+
+    /**
+     * Adds an authorization predicate.
+     */
+    public HttpAuthServiceBuilder add(Predicate<? super HttpHeaders> predicate) {
+        predicates.add(predicate);
+        return this;
+    }
+
+    /**
+     * Adds multiple authorization predicates.
+     */
+    public HttpAuthServiceBuilder add(Iterable<? extends Predicate<? super HttpHeaders>> predicates) {
+        this.predicates.addAll(Lists.newArrayList(predicates));
+        return this;
+    }
+
+    /**
+     * Adds an HTTP basic authorization predicate.
+     */
+    public HttpAuthServiceBuilder addBasicAuth(Predicate<? super BasicToken> predicate) {
+        this.predicates.add(new Predicate<HttpHeaders>() {
+            private final Function<HttpHeaders, BasicToken> extractor = AuthTokenExtractors.BASIC;
+
+            @Override
+            public boolean test(HttpHeaders headers) {
+                BasicToken token = extractor.apply(headers);
+                return token != null ? predicate.test(token) : false;
+            }
+        });
+        return this;
+    }
+
+    /**
+     * Adds an OAuth1a authorization predicate.
+     */
+    public HttpAuthServiceBuilder addOAuth1a(Predicate<? super OAuth1aToken> predicate) {
+        this.predicates.add(new Predicate<HttpHeaders>() {
+            private final Function<HttpHeaders, OAuth1aToken> extractor = AuthTokenExtractors.OAUTH1A;
+
+            @Override
+            public boolean test(HttpHeaders headers) {
+                OAuth1aToken token = extractor.apply(headers);
+                return token != null ? predicate.test(token) : false;
+            }
+        });
+        return this;
+    }
+
+    /**
+     * Adds an OAuth2 authorization predicate.
+     */
+    public HttpAuthServiceBuilder addOAuth2(Predicate<? super OAuth2Token> predicate) {
+        this.predicates.add(new Predicate<HttpHeaders>() {
+            private final Function<HttpHeaders, OAuth2Token> extractor = AuthTokenExtractors.OAUTH2;
+
+            @Override
+            public boolean test(HttpHeaders headers) {
+                OAuth2Token token = extractor.apply(headers);
+                return token != null ? predicate.test(token) : false;
+            }
+        });
+        return this;
+    }
+
+    /**
+     * Creates a new {@link HttpAuthService} instance with the given delegate and all of the authorization
+     * {@link Predicate}s.
+     */
+    public HttpAuthService build() {
+        return new HttpAuthServiceImpl(delegate, Iterables.toArray(predicates, Predicate.class));
+    }
+
+    /**
+     * Creates a new {@link HttpAuthService} {@link Service} decorator that supports all of the given
+     * authorization {@link Predicate}s.
+     */
+    public Function<Service<? super HttpRequest, ? extends HttpResponse>, HttpAuthService> newDecorator() {
```

Comment:
Other builders actually require the delegate in `newDecorator()` method. Could you make this builder has the same style of API? See `CorsServiceBuilder.build()` and `newDecorator()` for an example.